pkgs/build-support/shell-builder.nix

```diff
@@ -0,0 +1,41 @@
+{ lib, stdenv }:
+
+# A special kind of derivation that is only meant to be consumed by the
+# nix-shell.
+{
+  mergeInputs ? [], # a list of derivations whose inputs will be merged
+  buildInputs ? [],
+  nativeBuildInputs ? [],
+  propagatedBuildInputs ? [],
+  ...
+}@attrs:
+let
+  mergeInputs' = name:
+    let
+      op = item: sum: sum ++ item."${name}" or [];
+      nul = [];
+      list = [attrs] ++ mergeInputs;
+    in
+      lib.foldr op nul list;
+
+  rest =
+    builtins.removeAttrs
+      attrs
+      ["mergeInputs" "buildInputs" "nativeBuildInputs" "propagatedBuildInputs"];
+in
+
+stdenv.mkDerivation ({
+  name = "nix-shell";
+  phases = ["nobuildPhase"];
+
+  buildInputs = mergeInputs' "buildInputs";
+  nativeBuildInputs = mergeInputs' "nativeBuildInputs";
+  propagatedBuildInputs = mergeInputs' "propagatedBuildInputs";
```

Comment:
You are missing `propagatedNativeBuildInputs`.

Reply:
Thanks, I didn't know this existed, but makes sense. Fixed in b1118608258 which will be squashed later